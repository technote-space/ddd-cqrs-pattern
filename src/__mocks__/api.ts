import aspida from '@aspida/axios';
import Api from '@/web/shared/api';
import api from '^/pages/api/$api';

export class TestApi extends Api {
  public constructor(setUser?: () => void) {
    super(api(aspida()) as never, {
      setUser: setUser ?? (() => {
        //
      }),
    } as never);
  }
}
