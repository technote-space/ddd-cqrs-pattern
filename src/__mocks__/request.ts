import type { NextApiRequest } from 'next';

type Params<T> = {
  query?: Record<string, string>;
  body?: T;
  headers?: Record<string, string>;
};
export const createRequest = <T>(params?: Params<T>): NextApiRequest => ({
  query: params?.query ?? {},
  headers: params?.headers ?? {},
  body: params?.body,
}) as NextApiRequest;
