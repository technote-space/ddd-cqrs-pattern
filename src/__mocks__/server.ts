import type { RestHandler } from 'msw';
import type { DefaultRequestBody } from 'msw/lib/types/handlers/RequestHandler';
import type { RequestParams, RestRequest } from 'msw/lib/types/handlers/RestHandler';
import { rest } from 'msw';
import { setupServer } from 'msw/node';

export const mockNotionBaseUrl = 'https://example.com';
type Method = keyof typeof rest;
type Body = Record<string, any> | ((req: RestRequest<DefaultRequestBody, RequestParams>) => Record<string, any>);
type TestRequest = (req: RestRequest<DefaultRequestBody, RequestParams>) => void;
export const createHandler = (method: Method, path: string, status: number, jsonBody: Body, testRequest?: TestRequest): RestHandler => rest[method](path, (req, res, ctx) => {
  if (testRequest) testRequest(req);
  return res(
    ctx.status(status),
    ctx.json(typeof jsonBody === 'function' ? jsonBody(req) : jsonBody),
  );
});
export const createNotionHandler = (method: Method, path: string, status: number, jsonBody: Body, testRequest?: TestRequest): RestHandler => createHandler(method, `${mockNotionBaseUrl}/v1${path}`, status, jsonBody, testRequest);
export const useMockServer = (handlers: RestHandler[]) => {
  const server = setupServer(...handlers);
  beforeAll(() => server.listen({ onUnhandledRequest: 'error' }));
  afterEach(() => server.resetHandlers());
  afterAll(() => server.close());
};
