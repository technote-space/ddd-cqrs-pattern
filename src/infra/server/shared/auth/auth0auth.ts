import type IAuth from '$/server/shared/auth';
import type { AuthContents } from '$/server/shared/auth';
import type IEnv from '$/server/shared/env';
import axios from 'axios';
import { singleton, inject } from 'tsyringe';

@singleton()
export default class Auth0Auth implements IAuth {
  public constructor(
    @inject('IEnv') private env: IEnv,
  ) {
  }

  public async verify(token: string): Promise<null | AuthContents> {
    try {
      const response = await axios.get(`https://${this.env.getValue('NEXT_PUBLIC_AUTH0_DOMAIN')}/userinfo`, {
        headers: {
          Authorization: `Bearer ${token}`,
        },
      });

      const contents = await response.data as { sub: string };
      return {
        sub: contents.sub,
      };
    } catch (e) {
      return null;
    }
  }
}
