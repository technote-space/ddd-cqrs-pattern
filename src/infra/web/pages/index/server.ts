import type { IIndexPageProps, Props } from '$/web/pages';
import type { GetStaticPropsResult } from 'next';
import { singleton } from 'tsyringe';

@singleton()
export default class IndexPageProps implements IIndexPageProps {
  public async getStaticProps(): Promise<GetStaticPropsResult<Props>> {
    return {
      props: {},
      revalidate: undefined,
    };
  }
}
