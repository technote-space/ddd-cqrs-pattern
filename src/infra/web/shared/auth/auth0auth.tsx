import type IEnv from '$/server/shared/env';
import type { IApi } from '$/web/shared/api';
import type { IAuthContext, IAuth, StoreContext, UserResult, LogoutCallback } from '$/web/shared/auth';
import type { IContextProvider } from '$/web/shared/contextProvider';
import type { Dispatch, Reducer } from '$/web/shared/store';
import type { AppState } from '@auth0/auth0-react';
import { useAuth0 } from '@auth0/auth0-react';
import { Auth0Provider } from '@auth0/auth0-react';
import { createBrowserHistory } from 'history';
import { memo, PropsWithChildren, useCallback, useEffect, VFC } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { inject, singleton } from 'tsyringe';
import { useAddProcess, useDeleteProcess, useIsProcessRunning, useLoading } from '@/web/shared/loading';

@singleton()
export class AuthContext implements IAuthContext {
  private readonly SET_USER = 'SET_USER';

  public getKey(): string {
    return 'auth';
  }

  public getInitialState(): StoreContext {
    return { user: { isLoggedIn: false } };
  }

  public getReducerMapObject(): Record<string, Reducer<StoreContext>> {
    return {
      [this.SET_USER]: (store, action) => ({ ...store, user: action.user }),
    };
  }

  public persistTargets(): Array<keyof StoreContext> {
    return ['user'];
  }

  public useUser(): UserResult {
    /* istanbul ignore next */
    const selector = (state: { auth: StoreContext }) => state.auth.user;
    return useSelector(selector);
  }

  public setUser(dispatch: Dispatch, user: UserResult) {
    dispatch({ type: this.SET_USER, user });
  }
}

export type Auth0Config = {
  domain: string;
  clientId: string;
}

@singleton()
export class Auth0Auth implements IAuth {
  private _syncLoading = false;

  public constructor(
    @inject('IAuthContext') private authContext: IAuthContext,
    @inject('IApi') private api: IApi,
  ) {
  }

  public useUser(): UserResult {
    const { isLoading, isAuthenticated, error, getAccessTokenSilently, loginWithRedirect } = useAuth0();
    const user = this.authContext.useUser();
    const dispatch = useDispatch();
    const withLoading = useLoading();
    const addProcess = useAddProcess();
    const deleteProcess = useDeleteProcess();
    const isProcessRunning = useIsProcessRunning();
    const caller = this.api.useCaller();

    useEffect(() => {
      if (isLoading) {
        addProcess('loadingAuth0', '認証確認中...');
      } else {
        deleteProcess('loadingAuth0');
      }
    }, [isLoading, addProcess, deleteProcess]);

    useEffect(() => {
      if (isLoading) {
        return;
      }

      // auth0 にログインしていない かつ ログイン状態の場合はログアウト
      if (!isAuthenticated && user.isLoggedIn) {
        this.authContext.setUser(dispatch, {
          isLoggedIn: false,
        });
      }

      // auth0 にログインしていてログイン状態でない場合はログイン
      if (isAuthenticated && !user.isLoggedIn && !isProcessRunning('login') && !this._syncLoading) {
        this._syncLoading = true;
        (async () => {
          await withLoading(async () => {
            const { authorization } = await caller(async client => client.login.$post({ body: { token: await getAccessTokenSilently() } }), { authorization: '' });
            if (authorization) {
              this.authContext.setUser(dispatch, {
                user: { authorization },
                isLoggedIn: true,
              });
            }
          }, 'ログイン中...', 'login');
          this._syncLoading = false;
        })();
      }

      // auth0 にログインしていない場合はログインページにリダイレクト
      if (!isAuthenticated && !error) {
        loginWithRedirect({
          appState: { page: window.location.pathname },
        }).then();
      }
    }, [
      dispatch, isLoading, error, isAuthenticated, user.isLoggedIn,
      caller, getAccessTokenSilently, loginWithRedirect, withLoading, isProcessRunning,
    ]);

    return user;
  }

  public useLogout(): LogoutCallback {
    const { logout } = useAuth0();
    return useCallback(async () => {
      logout();
    }, [logout]);
  }
}

@singleton()
export class Auth0ContextProvider implements IContextProvider {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  private readonly __provider: VFC<PropsWithChildren<any>>;

  public constructor(
    @inject('IEnv') private env: IEnv,
    @inject('auth0Config') private config: Auth0Config,
  ) {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    const provider = memo(({ children }: PropsWithChildren<any>) => {
      return <Auth0Provider
        redirectUri={this.getRedirectUri()}
        onRedirectCallback={Auth0ContextProvider.onRedirectCallback}
        {...this.config}
      >
        {children}
      </Auth0Provider>;
    });
    provider.displayName = 'AuthProvider';
    this.__provider = provider;
  }

  private getRedirectUri() {
    return `${window.location.origin}${this.env.getValue('BASE_PATH', '')}`;
  }

  /* istanbul ignore next */
  private static onRedirectCallback(appState: AppState) {
    if (appState?.page) {
      const history = createBrowserHistory();
      history.replace(appState.page);
    }
  }

  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  public getProvider(): VFC<PropsWithChildren<any>> {
    return this.__provider;
  }
}
