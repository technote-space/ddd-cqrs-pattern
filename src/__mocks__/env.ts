import type IEnv from '$/server/shared/env';
import { mockNotionBaseUrl } from '^/__mocks__/server';

export default class TestEnv implements IEnv {
  public constructor(private env: Record<string, string>) {
    this.env.NOTION_BASE_URL = mockNotionBaseUrl;
  }

  public isValid(key: string): boolean {
    return key in this.env && this.env[key] !== undefined;
  }

  public getValue(key: string, defaultValue?: string): string | never {
    if (!(key in this.env)) {
      if (defaultValue !== undefined) {
        return defaultValue;
      }

      throw Error('環境変数が設定されていません: ' + key);
    }

    const env = this.env[key];
    if (env === undefined) {
      if (defaultValue !== undefined) {
        return defaultValue;
      }

      throw Error('環境変数が設定されていません: ' + key);
    }

    return env;
  }
}
