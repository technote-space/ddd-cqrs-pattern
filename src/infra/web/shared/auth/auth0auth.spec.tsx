import type { IAuthContext, StoreContext, UserResult } from '$/web/shared/auth';
import type { ILoading, ILoadingContext } from '$/web/shared/loading';
import type { Auth0ContextInterface } from '@auth0/auth0-react/dist/auth0-context';
import * as auth0 from '@auth0/auth0-react';
import { renderHook } from '@testing-library/react-hooks';
import * as redux from 'react-redux';
import renderer from 'react-test-renderer';
import TestEnv from '^/__mocks__/env';
import { createLocalHandler, useMockServer } from '^/__mocks__/server';
import { Auth0Auth, AuthContext } from './auth0auth';

jest.mock('react-redux');

describe('AuthContext', () => {
  const context = new AuthContext();

  it('key が auth', () => {
    expect(context.getKey()).toBe('auth');
  });

  it('user の初期値は isLoggedIn = false', () => {
    expect(context.getInitialState()).toEqual({ user: { isLoggedIn: false } });
  });

  it('SET_USER で user 情報が設定', () => {
    const mapper = context.getReducerMapObject();
    const store = { user: { isLoggedIn: false } } as StoreContext;
    const user = { isLoggedIn: true, user: { authorization: 'authorization' } };

    expect(mapper).toHaveProperty('SET_USER');
    expect(typeof mapper['SET_USER']).toBe('function');
    expect(mapper['SET_USER'](store, { type: 'SET_USER', user })).toEqual({ user });
    expect(mapper['SET_USER'](undefined, { type: 'SET_USER', user })).toEqual({ user });
  });

  it('永続化の対象はなし', () => {
    expect(context.persistTargets()).toEqual([]);
  });

  it('user の値を取得する', () => {
    jest.spyOn(redux, 'useSelector').mockImplementation(() => ({ isLoggedIn: false }));

    expect(context.useUser()).toEqual({ isLoggedIn: false });
  });

  it('user の値を更新する', () => {
    const mockDispatch = jest.fn();
    const user = { isLoggedIn: false } as UserResult;

    context.setUser(mockDispatch, user);

    expect(mockDispatch).toBeCalledWith({ type: 'SET_USER', user });
  });
});

describe('Auth0Auth', () => {
  it('Auth0がローディング中の場合にLoadingContextにローディング中のプロセス登録される', () => {
    jest.spyOn(auth0, 'useAuth0').mockReturnValue({
      isLoading: true,
      isAuthenticated: false,
      error: undefined,
      getAccessTokenSilently: jest.fn(),
      loginWithRedirect: jest.fn(),
    } as never as Auth0ContextInterface);

    const mockAdd = jest.fn();
    const mockDelete = jest.fn();
    const auth = new Auth0Auth(
      { useUser: () => ({ isLoggedIn: false }) } as never as IAuthContext,
      { useProcess: () => [], add: mockAdd, delete: mockDelete } as never as ILoadingContext,
      { useLoading: jest.fn() } as never as ILoading,
      new TestEnv({}),
      { domain: '', clientId: '' },
    );

    const result = renderHook(() => auth.useUser()).result;

    expect(result.current).toEqual({ isLoggedIn: false });
    expect(mockAdd).toBeCalledTimes(1);
    expect(mockDelete).not.toBeCalled();
  });

  describe('Auth0のロードが完了', () => {
    useMockServer([
      createLocalHandler('post', '/login', 200, {
        authorization: 'token',
      }),
    ]);

    describe('Auth0にログイン成功', () => {
      it('未ログイン状態の場合でまだログインを試みていない場合はログイン', async () => {
        jest.spyOn(auth0, 'useAuth0').mockReturnValue({
          isLoading: false,
          isAuthenticated: true,
          error: undefined,
          getAccessTokenSilently: jest.fn(),
          loginWithRedirect: jest.fn(),
        } as never as Auth0ContextInterface);

        const mockAdd = jest.fn();
        const mockDelete = jest.fn();
        const mockSetUser = jest.fn();
        const mockWithLoading = jest.fn(callback => callback());
        const auth = new Auth0Auth(
          { useUser: () => ({ isLoggedIn: false }), setUser: mockSetUser } as never as IAuthContext,
          { useProcess: () => [], add: mockAdd, delete: mockDelete } as never as ILoadingContext,
          { useLoading: () => mockWithLoading, isProcessRunning: () => false } as never as ILoading,
          new TestEnv({}),
          { domain: '', clientId: '' },
        );

        const { result, waitFor } = renderHook(() => auth.useUser());

        expect(result.current).toEqual({ isLoggedIn: false });
        expect(mockAdd).not.toBeCalled();
        expect(mockDelete).toBeCalledTimes(1);

        expect(mockWithLoading).toBeCalledTimes(1);
        await waitFor(() => expect(mockSetUser).toBeCalledTimes(1));
        expect(mockSetUser).toBeCalledWith(undefined, { user: { authorization: 'token' }, isLoggedIn: true });
      });

      it('未ログイン状態の場合ですでにログインを試みている場合はログインしない', () => {
        jest.spyOn(auth0, 'useAuth0').mockReturnValue({
          isLoading: false,
          isAuthenticated: true,
          error: undefined,
          getAccessTokenSilently: jest.fn(),
          loginWithRedirect: jest.fn(),
        } as never as Auth0ContextInterface);

        const mockAdd = jest.fn();
        const mockDelete = jest.fn();
        const mockSetUser = jest.fn();
        const mockWithLoading = jest.fn(callback => callback());
        const auth = new Auth0Auth(
          { useUser: () => ({ isLoggedIn: false }), setUser: mockSetUser } as never as IAuthContext,
          { useProcess: () => [{ id: 'login' }], add: mockAdd, delete: mockDelete } as never as ILoadingContext,
          { useLoading: () => mockWithLoading, isProcessRunning: () => true } as never as ILoading,
          new TestEnv({}),
          { domain: '', clientId: '' },
        );

        const { result } = renderHook(() => auth.useUser());

        expect(result.current).toEqual({ isLoggedIn: false });
        expect(mockAdd).not.toBeCalled();
        expect(mockDelete).toBeCalledTimes(1);

        expect(mockWithLoading).not.toBeCalled();
      });
    });

    describe('Auth0にログイン失敗', () => {
      it('ログイン状態の場合はログアウト', () => {
        const mockLoginWithRedirect = jest.fn();
        jest.spyOn(auth0, 'useAuth0').mockReturnValue({
          isLoading: false,
          isAuthenticated: false,
          error: new Error(),
          getAccessTokenSilently: jest.fn(),
          loginWithRedirect: mockLoginWithRedirect,
        } as never as Auth0ContextInterface);

        const mockAdd = jest.fn();
        const mockDelete = jest.fn();
        const mockSetUser = jest.fn();
        const mockWithLoading = jest.fn(callback => callback());
        const auth = new Auth0Auth(
          {
            useUser: () => ({ isLoggedIn: true, user: { authorization: 'token' } }),
            setUser: mockSetUser,
          } as never as IAuthContext,
          { useProcess: () => [], add: mockAdd, delete: mockDelete } as never as ILoadingContext,
          { useLoading: () => mockWithLoading, isProcessRunning: () => false } as never as ILoading,
          new TestEnv({}),
          { domain: '', clientId: '' },
        );

        const { result } = renderHook(() => auth.useUser());

        expect(result.current).toEqual({ isLoggedIn: true, user: { authorization: 'token' } });
        expect(mockAdd).not.toBeCalled();
        expect(mockDelete).toBeCalledTimes(1);

        expect(mockWithLoading).not.toBeCalled();
        expect(mockSetUser).toBeCalledWith(undefined, { isLoggedIn: false });
        expect(mockLoginWithRedirect).not.toBeCalled();
      });

      it('エラーではない場合はログインページにリダイレクト', () => {
        const mockLoginWithRedirect = jest.fn(() => Promise.resolve());
        jest.spyOn(auth0, 'useAuth0').mockReturnValue({
          isLoading: false,
          isAuthenticated: false,
          error: undefined,
          getAccessTokenSilently: jest.fn(),
          loginWithRedirect: mockLoginWithRedirect,
        } as never as Auth0ContextInterface);

        const mockAdd = jest.fn();
        const mockDelete = jest.fn();
        const mockSetUser = jest.fn();
        const mockWithLoading = jest.fn(callback => callback());
        const auth = new Auth0Auth(
          {
            useUser: () => ({ isLoggedIn: false }),
            setUser: mockSetUser,
          } as never as IAuthContext,
          { useProcess: () => [], add: mockAdd, delete: mockDelete } as never as ILoadingContext,
          { useLoading: () => mockWithLoading, isProcessRunning: () => false } as never as ILoading,
          new TestEnv({}),
          { domain: '', clientId: '' },
        );

        const { result } = renderHook(() => auth.useUser());

        expect(result.current).toEqual({ isLoggedIn: false });
        expect(mockAdd).not.toBeCalled();
        expect(mockDelete).toBeCalledTimes(1);

        expect(mockWithLoading).not.toBeCalled();
        expect(mockSetUser).not.toBeCalled();
        expect(mockLoginWithRedirect).toBeCalledTimes(1);
      });
    });
  });

  it('ログアウトが呼ばれる', () => {
    const mockLogout = jest.fn();
    jest.spyOn(auth0, 'useAuth0').mockReturnValue({
      logout: mockLogout,
    } as never as Auth0ContextInterface);

    const auth = new Auth0Auth(
      {} as never as IAuthContext,
      {} as never as ILoadingContext,
      {} as never as ILoading,
      new TestEnv({}),
      { domain: '', clientId: '' },
    );

    const result = renderHook(() => auth.useLogout()).result;
    result.current();

    expect(mockLogout).toBeCalledTimes(1);
  });

  it('Authのコンテキストプロバイダー', () => {
    const auth = new Auth0Auth(
      {} as never as IAuthContext,
      {} as never as ILoadingContext,
      {} as never as ILoading,
      new TestEnv({}),
      { domain: '', clientId: '' },
    );

    const Provider = auth.getAuthProvider();
    const tree = renderer.create(<Provider>test</Provider>);
    expect(tree).toMatchSnapshot();
  });
});
