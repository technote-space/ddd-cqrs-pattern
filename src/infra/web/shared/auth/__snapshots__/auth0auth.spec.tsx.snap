// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Auth0Auth Authのコンテキストプロバイダー 1`] = `"test"`;
